bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,395 @@
+        final String actionUrl = "\"http://purenetworks.com/HNAP1/" + action + "\"";
+        request.header("SOAPAction", actionUrl);
+        if (authentication != null) {
+            request.header("HNAP_AUTH", computeHnapAuth(actionUrl, authentication.privateKey));
+            request.header("Cookie", "uid=" + authentication.cookie);
+        }
+        try {
+            request.content(new BytesContentProvider(encodeSOAPMessage(message)));
+            final ContentResponse response = request.send();
+            final byte[] content = response.getContent();
+            return DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(new ByteArrayInputStream(content));
+        } catch (InterruptedException | TimeoutException | ExecutionException | SAXException | IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            throw new CommunicationException("Post request failed: " + e.getMessage());
+        } catch (ParserConfigurationException e) {
+            logger.error("Failed to create XML parser.");
+            updateStatus(ThingStatus.OFFLINE);
+            return null;
+        }
+    }
+
+    private byte[] encodeSOAPMessage(final SOAPMessage message) {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        try {
+            message.writeTo(stream);
+        } catch (SOAPException | IOException e) {
+            logger.error("Unexpected internal Error.");
+            updateStatus(ThingStatus.OFFLINE);
+        }
+        return stream.toByteArray();
+    }
+
+    private String getElementByTagName(final Document document, final String name) throws CommunicationException {
+        NodeList items = document.getDocumentElement().getElementsByTagName(name);
+        final int n = items.getLength();
+        if (n != 1) {
+            logger.error("Failed to get element with tag name '{}'.", name);
+            authentication = null; // force re-authentication.
+            throw new CommunicationException(
+                    "Expected exactly one node with name '" + name + "', but got " + n + " many.");
+        } else {
+            return items.item(0).getTextContent();
+        }
+    }
+
+    private String computeHnapAuth(final String url, final String privateKey) {
+        if (url.endsWith("/Login")) {
+            return "\"" + privateKey + "\"";
+        } else {
+            final String timeStamp = Long.toString(Instant.now().getEpochSecond() / (int) 10e6);
+            return calculateHMAC(timeStamp + url, privateKey) + " " + timeStamp;
+        }
+    }
+
+    private static final String HMAC_MD5 = "HmacMD5";
+
+    private static String toHexString(byte[] bytes) {
+        Formatter formatter = new Formatter();
+        for (byte b : bytes) {
+            formatter.format("%02X", b);
+        }
+        final String hmac = formatter.toString();
+        formatter.close();
+        return hmac;
+    }
```

Comment:
You could use the built-in converter: https://www.openhab.org/javadoc/v2.5/org/eclipse/smarthome/core/util/hexutils